include/simdjson/padded_string.h

```diff
@@ -10,34 +10,92 @@ namespace simdjson {
 // "length" bytes safely. if you must provide a pointer to some data, create it
 // with this function: length is the max. size in bytes of the string caller is
 // responsible to free the memory (free(...))
-char *allocate_padded_buffer(size_t length);
+inline char *allocate_padded_buffer(size_t length) noexcept {
+
+#ifndef NDEBUG
+  if (length < 1) {
+    errno = EINVAL;
+    perror("simdjson::allocate_padded_buffer() length argument is less than 1");
+    return nullptr;
+  }
+#endif // NDEBUG
+
+  // we could do a simple malloc
+  // return (char *) malloc(length + SIMDJSON_PADDING);
+  // However, we might as well align to cache lines...
+  size_t totalpaddedlength = length + SIMDJSON_PADDING;
+  char *padded_buffer = aligned_malloc_char(64, totalpaddedlength);
+  memset(padded_buffer + length, 0, totalpaddedlength - length);
+  return padded_buffer;
+} // allocate_padded_buffer
 
 // Simple string with padded allocation.
 // We deliberately forbid copies, users should rely on swap or move
 // constructors.
-class padded_string {
-public:
+struct padded_string final {
+
   explicit padded_string() noexcept : viable_size(0), data_ptr(nullptr) {}
+
   explicit padded_string(size_t length) noexcept
       : viable_size(length), data_ptr(allocate_padded_buffer(length)) {
 
     if (data_ptr != nullptr)
       data_ptr[length] = '\0'; // easier when you need a c_str
+#ifndef NDEBUG
+    else {
+      errno = EINVAL;
+      perror("simdjson::padded_string() length argument is less than 1");
+      // what to do here? exit() ?
+    }
+#endif
   }
+
   explicit padded_string(char *data, size_t length) noexcept
       : viable_size(length), data_ptr(allocate_padded_buffer(length)) {
     if (data_ptr != nullptr) {
       memcpy(data_ptr, data, length);
       data_ptr[length] = '\0'; // easier when you need a c_str
     }
+#ifndef NDEBUG
+    else {
+      errno = EINVAL;
+      perror("simdjson::padded_string() data argument is null");
+      // what to do here? exit() ?
+    }
+#endif
   }
-  padded_string(std::string s) noexcept
+
+  // note: do not have std::string arguments passed by value
+  padded_string(std::string const &s) noexcept
```

Comment:
I think that elsewhere in the library we put the const qualifiers before the type. Unless you have a good reason to break the style, I recommend moving the const qualifier before std::string.